Blog: Collector Kafka Improvements (#9924)
open-telemetry/opentelemetry.io: content/en/blog/2026/collector-kafka-improvements.md

Author: Please feel free to provide any feedback as suggestions :D

Author: I've kept ordering of code owners / authors as alphabetical order based on given name, if there is a preferred order then please let me know :)

---
title: OpenTelemetry Collector Kafka Components Improvements
linkTitle: Collector Kafka Improvements
sig: Collector SIG
date: {{ dataFormat "2026-01-02" .Date }}
author: >-
  [Andrew Wilkins](https://github.com/axw) (Elastic),
  [Paulo Dias](https://github.com/paulojmdias) (Five9),
  [Pavol Loffay](https://github.com/pavolloffay) (Red Hat),
  [Sean Marciniak](https://github.com/MovieStoreGuy) (Splunk)
draft: true
body_class: otel-with-contributions-from
---

The code owners for the Kafka components have been looking at improving the
maintainability of each component, ensuring consistent configuration, and
performance improvements where possible. Thanks to all who have contributed
to this effort, as of `v0.152.0` the Collector Contrib now ships with all
improvements by default.

## Background

> 
> Historically, each of these Kafka components were updated independently and were

Reviewer: ```suggestion

OpenTelemetry Collector Contrib contains a variety of Kafka-related components:
  - [exporter/kafka](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/kafkaexporter) for producing data to Kafka topics.
  - [receiver/kafka](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/kafkareceiver) for receiving data stored in Kafka topics.
  - [receiver/kafkametrics](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/kafkametricsreceiver) for scraping metrics from Kafka brokers.
  - [extension/kafkatopics_observer](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/kafkatopicsobserver) for reporting Kafka topics to the [Receiver Creator](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/receivercreator/README.md).

Historically, each of these Kafka components was updated independently and
```

> optimized for their specific use case. This meant that some configuration that

Reviewer: ```suggestion
optimized for its specific use case. This meant that some configuration that
```

should have been adopted for all impacted components was only added for that
specific component. Due to this, any potential consolidation efforts, or just
general maintenance required took a lot more effort than it should have.
This was definitely felt by our users through constant configuration changes,
changes in behaviour, and inconsistent expectations — something the code owners
set out to fix.

> ## Improvements

Author: I would like to find all the issues that we've closed and share them here but I feel that might lengthy, so I am open to feedback.

> 

Reviewer: [Marc Lopez Rubio](https://github.com/marclop) introduced the use of the [franz-go](https://github.com/twmb/franz-go) Kafka client to exporter/kafka (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40364) and receiver/kafka (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40716), which greatly improved performance. Micro-benchmarks show the exporter taking ~37% less time, ~59% less memory, and ~46% fewer allocations. The receiver benchmarks show throughput increasing by several orders of magnitude.

In addition to these very welcome performance improvements, moving to franz-go also brought support for regex topic consuming to the receiver. This was previously enabled by the use of the receiver creator and extension/kafkatopics_observer, but this approach had a significant issue: every topic would have had its own Kafka client, connections to the Kafka brokers, metadata, etc. -- this quickly adds up when consuming from many topics.

## Acknowledgements

These changes could not have been done without the direction of:

- [Andrew Wilkins](https://github.com/axw) (Elastic),
- [Paulo Dias](https://github.com/paulojmdias) (Five9),
- [Pavol Loffay](https://github.com/pavolloffay) (Red Hat),
- [Sean Marciniak](https://github.com/MovieStoreGuy) (Splunk)

Furthermore, we would also like to acknowledge the help we have received from 
[Travis Bischel](https://github.com/twmb) as we've made new changes, needed new

> features within the Kafka SDK, they have been amazing to collaborate with as

Reviewer: ```suggestion
features within the [franz-go](https://github.com/twmb/franz-go) Kafka client, they have been amazing to collaborate with as
```
